WebSocket stubbing (#577)
softwaremill/sttp: docs/testing.md

val response2 = basicRequest.post(uri"http://api.internal/b").send()
// response2 will be whatever a "real" network call to api.internal/b returns
```

## Testing WebSockets

> Stub methods `whenRequestMatches` and `whenAnyRequest` also allow to specify behavior for `openWebsocket` call when used together with `thenRespondWebSocket` or `thenHandleOpenWebSocket`. `thenRespondWebSocket` is intended for cases when sttp supplied

Reviewer: Unneeded line break at the end of this line (i.e. lines 204 and 205 should be one line)

Author: :+1: I've edited my changes to apply line breaks before 121 column. Markdown should render single white space in same way, regardless it's line break or space.

Reviewer: Excellent, I didn't know that

> `WebSocketHandler` is used, that are cases where users code manipulates `WebSocket[F]` to implement logic.

Reviewer: ```suggestion
`WebSocketHandler` is used, i.e. the cases where the user's code manipulates `WebSocket[F]` to implement logic.
```

Reviewer: (if I understand it correctly)

This way `thenRespondWebSocket` can be used to return test double `WebSocket`.

Example:
```scala
val testWebSocket: WebSocket[Task] = ???

implicit val testingBackend =
  AsyncHttpClientZioBackend.stub
    .whenAnyRequest
    .thenRespondWebSocket(Headers(List.empty), testWebSocket)

for {
  handler <- ZioWebSocketHandler() // using sttp supplied handler
  openResponse <- testingBackend.openWebsocket(basicRequest.get(uri"wss://some.uri"), handler)
  webSocket = openResponse.result // interaction with WebSocket is interesting in this case
  message <- webSocket.receive
} yield message
```

## WebSocketStub

> Implementation of `WebSocket` stub is available. It wraps state machine that can be used to simulate simple WebSockets interactions.

Reviewer: ```suggestion
There is an implementation of `WebSocket` available. It wraps a state machine that can be used to simulate simple WebSockets interactions.
```

It allows to set initial responses for `receive` calls as well as logic to add further messages in reaction to `send` calls.

> `SttpBackendStub` has special API that accepts `WebSocketStub` and builds it to `WebSocket` instance.

Reviewer: ```suggestion
`SttpBackendStub` has a special API that accepts a `WebSocketStub` and builds a `WebSocket` instance out of it.
```

Reviewer: (if I understand that correctly)

Example:
```
val backend = SttpBackendStub.synchronous[Identity]
val webSocketStub = WebSocketStub
  .withInitialIncoming(
    List(WebSocketFrame.text("Hello from the server!"))
  )
  .thenRespondS(0) {
    case (counter, tf: WebSocketFrame.Text) => (counter + 1, List(WebSocketFrame.text(s"echo: ${tf.payload}")))
    case (counter, _)                       => (counter, List.empty)
  }
backend.whenAnyRequest.thenRespondWebSocket(Headers(List.empty), webSocketStub)
```

> There is a possiblity to add error responses as well, if it's not enough,

Reviewer: There's a lot of extra line breaks in the middle of the sentences, throughout this file - they should be removed.

> then using own implementation of `WebSocket` is recommended.

Reviewer: ```suggestion
then using a custom implementation of `WebSocket` is recommended.
```

## Using handler when testing WebSockets

> Akka backend works a bit differently in regards to WebSockets. In it's case user defines

Reviewer: ```suggestion
Akka backend works a bit differently in regard to WebSockets. In this case, the user defines
```

> `Flow[Message, Message, Mat]` to define WebSocket client behavior instead of working with `WebSocket` instance.

Reviewer: ```suggestion
a `Flow[Message, Message, Mat]` to define WebSocket client behavior instead of working with `WebSocket` instance.
```

`thenHandleOpenWebSocket` method allows to use given flow to simulate server side in tests.

Example:
```scala
// it should say Hi! and 42
val behaviorToTest: Flow[Message, Message, Future[Done]] = ???
// setup test double
var received = List.empty[String]

> val testFlow: Flow[Message, Message, Future[Done]] => Future[Done] = clientFlow => {

Reviewer 2: `serverFlow`?
